README.md
deeper:Deep Ensemble for Environmental Predictor
================
Alven Yu
2022/03/03

-   [The framework of DEML model](#the-framework-of-deml-model)
    -   [\* Step 1: establish the base
        models](#-step-1-establish-the-base-models)
    -   [\* Step 2: stack the meta
        models](#-step-2-stack-the-meta-models)
    -   [\* Step 3: prediction based on the new data
        set](#-step-3-prediction-based-on-the-new-data-set)
-   [Installation](#installation)
-   [Example](#example)
    -   [1. Data preparation](#1-data-preparation)
        -   [Load the example data and separate to training and testing
            dataset](#load-the-example-data-and-separate-to-training-and-testing-dataset)
        -   [Identify the dependence and independence
            variables](#identify-the-dependence-and-independence-variables)
    -   [2.The models can be used](#2the-models-can-be-used)
    -   [3.Model Building](#3model-building)
    -   [3.1 Tuning the paramaters of a base
        model](#31-tuning-the-paramaters-of-a-base-model)
    -   [3.2 Establish DEML model](#32-establish-deml-model)
        -   [3.2.1 Training the (base) ensemble
            model](#321-training-the-base-ensemble-model)
        -   [3.2.2 Training the (base) ensemble model with parallel
            computing](#322-training-the-base-ensemble-model-with-parallel-computing)
    -   [3.3 Stacked meta models](#33-stacked-meta-models)
        -   [3.3.1 Stacked meta models based on trained base
            models](#331-stacked-meta-models-based-on-trained-base-models)
        -   [3.3.2 Stacked meta models from
            scratch](#332-stacked-meta-models-from-scratch)
        -   [3.3.3 Stacked meta models with paralleling
            computing](#333-stacked-meta-models-with-paralleling-computing)
    -   [4. Cross validation with paralleling
        computing](#4-cross-validation-with-paralleling-computing)
        -   [4.1 Cross validation with base
            models](#41-cross-validation-with-base-models)
        -   [4.2 Cross validation with stacked meta
            models](#42-cross-validation-with-stacked-meta-models)
    -   [5. Assess and plot the results](#5-assess-and-plot-the-results)
    -   [Citation](#citation)

<!-- badges: start -->

[![Codecov test
coverage](https://codecov.io/gh/Alven8816/deeper_1.0.0/branch/master/graph/badge.svg)](https://codecov.io/gh/Alven8816/deeper_1.0.0?branch=master)
[![R-CMD-check](https://github.com/Alven8816/deeper_1.0.0/workflows/R-CMD-check/badge.svg)](https://github.com/Alven8816/deeper_1.0.0/actions)
<!-- badges: end -->

## The framework of DEML model

The goal of deeper is to develop a multiple-level deep ensemble machine
learning framework (DEML) for improving the estimation of air pollution
concentrations.

![framework](./demlfigurev3.png)

The DEML framework proposed in this study is a three-level stacked
ensemble approach. It is based on the SuperLearner(SL) ensemble
algorithm (Naimi and Balzer 2018; Polley and Van Der Laan 2010; Van der
Laan et al. 2007) introduced in the neural network hierarchy structure.
Figure above illustrates the overall training procedure of our DEML
algorithm.

Our newly build R package mainly include 3 steps:

### \* Step 1: establish the base models

Using predictModel() or predictModel\_parallel() to establish the base
models. A tuningModel() function can be used to tuning the parameters to
get the best single base model.

### \* Step 2: stack the meta models

We use stack\_ensemble(),stack\_ensemble.fit(), or
stack\_ensemble\_parallel() function to stack the meta models to get a
DEML model.

### \* Step 3: prediction based on the new data set

After establishment of DEML model, the predict() can be used predict the
unseen data set.

To assess the performance of the models, assess.plot() can be used by
comparing the original observations (y in test set) and the prediction.
The assess.plot() also return the point scatter plot.

The other functions including CV.predictModel() (referred from
SuperLearner), CV.predictModel\_parallel(),and
CV.stack\_ensemble\_parallel() could be used to conduct the external
cross validation to assess the base models and DEML models for each
fold.

**Note: DEML could not directly deal with missing values and missing
value imputation technologies is recommended prior to the use of the
DEML model.**

------------------------------------------------------------------------

## Installation

You can install the developing version of deeper from
[github](https://github.com/Alven8816/deeper) with:

``` r
library(devtools)
install_github("Alven8816/deeper")
```

------------------------------------------------------------------------

# Example

This is a basic example which shows you how to use deeper:

## 1. Data preparation

### Load the example data and separate to training and testing dataset

``` r
library(deeper)

## obtain the example data
data("envir_example")
```

``` r
set.seed(1234)
size <-
  caret::createDataPartition(y = envir_example$PM2.5, p = 0.8, list = FALSE)
trainset <- envir_example[size, ]
testset <- envir_example[-size, ]
```

### Identify the dependence and independence variables

``` r
y <- c("PM2.5")
x <- colnames(envir_example[-c(1, 6)]) # except "date" and "PM2.5"
```

## 2.The models can be used

``` r
# we need to select the algorithm previously
SuperLearner::listWrappers()
data(model_list) # get more details about the algorithm
```

In the ‘type’ column, “R”: can be used for regression or
classification;“N”: can be used for regression but variables requre to
be numeric style; “C”: just used in classification.

## 3.Model Building

## 3.1 Tuning the paramaters of a base model

``` r
#method 1: using deeper package function
ranger <-
  tuningModel(
    basemodel  = 'SL.ranger',
    params = list(num.trees = 100),
    tune = list(mtry = c(1, 3, 7))
  )

#method 2: employ the Superlearner package "create.Learner" method
#ranger <- SuperLearner::create.Learner(base_learner = 'SL.ranger',params = list(num.trees = 1000),tune = list(mtry = c(1,3,7)))

#method 3: create a function by oneself as follow
#ranger <- function(...){SL.ranger(...,num.trees = 1000)}
```

## 3.2 Establish DEML model

### 3.2.1 Training the (base) ensemble model

``` r
#training the model with new dataset
#predict the model with several methods(here we choose XGBoost and Random Forest models)

pred_m1 <-
  predictModel(
    Y = trainset[, y],
    X = trainset[, x],
    newX = testset[, x],
    base_model = c("SL.xgboost", ranger),
    cvControl = list(V = 5)
  )

# return the newX prediction results
# get the ensemble model(SL) result
head(pred_m1$ensemble_pre)
# get the single base model results
head(pred_m1$single_pre)
```

The results show the weight, R2, and the root-mean-square error (RMSE)
of each model. “ranger\_1”,“ranger\_2”,“ranger\_3” note the Random
Forest model with parameters mtry = 1,3,7 separately.

The results show that mtry = 5 is the best RF model for this
prediction.This method could be used to tune the suitable parameters for
algorithms.

``` r
# training the model with separate new data set

pred_m2 <-
  predictModel(
    Y = trainset[, y],
    X = trainset[, x],
    base_model = c("SL.xgboost", ranger),
    cvControl = list(V = 5)
  )

# predict the new dataset
pred_m2_new <- predict(object = pred_m2, newX = testset[, x])
```

### 3.2.2 Training the (base) ensemble model with parallel computing

``` r
## conduct the spatial CV

# Create a list with 7 (folds) elements (each element contains index of rows to be considered on each fold)

indices <-
  CAST::CreateSpacetimeFolds(trainset, spacevar = "code", k = 7)

# Rows of validation set on each fold

v_raw <- indices$indexOut

names(v_raw) <- seq(1:7)

pred_m3 <- predictModel_parallel(
  Y = trainset[, y],
  X = trainset[, x],
  base_model = c("SL.xgboost", ranger),
  cvControl = list(V = length(v_raw), validRows = v_raw),
  number_cores = 4,
  seed = 1
)
## when number_cores is missing, it will indicate user to set one based on the operation system.

# pred_m3 <- predictModel_parallel(
#     Y = trainset[,y],
#     X = trainset[,x],
#     base_model = c("SL.xgboost",ranger),
#     cvControl = list(V = length(v_raw), validRows = v_raw),
#     seed = 1
#   )

#You have 8 cpu cores, How many cpu core you want to use:
# type the number to continue the process.


# prediction
pred_m3_new <- predict(object = pred_m3, newX = testset[, x])
```

## 3.3 Stacked meta models

Our DEML model includes several meta-models based on the trained base
models’ results. Different from other ensemble (stacking method), DEML
model allow to select several meta-models and run them parallelly and
ensemble all the meta-models with the optimal weights.(This step is
optinal for analysis).

### 3.3.1 Stacked meta models based on trained base models

``` r
#Do not include original feature
# object include newX dataset
pre_stack_1 <-
  stack_ensemble(
    object = pred_m1,
    meta_model = c("SL.ranger", "SL.xgboost", "SL.lm"),
    original_feature = FALSE
  )


# object do not include newX dataset
pre_stack_2 <-
  stack_ensemble(
    object = pred_m2,
    meta_model = c("SL.ranger", "SL.xgboost", "SL.lm"),
    original_feature = FALSE,
    X = testset[, x]
  )

# prediction
pred_stack_2_new <- predict(object = pre_stack_2, newX = testset[, x])


#Do include original feature
#object do not include newX dataset
pre_stack_3 <-
  stack_ensemble(
    object = pred_m2,
    meta_model = c("SL.ranger", "SL.xgboost", "SL.lm"),
    original_feature = TRUE,
    X = trainset[, x]
  )

# prediction
pred_stack_3_new <- predict(object = pre_stack_2, newX = testset[, x])

#object include newX dataset
pre_stack_4 <-
  stack_ensemble(
    object = pred_m1,
    meta_model = c("SL.ranger", "SL.xgboost", "SL.lm"),
    original_feature = TRUE,
    X = trainset[, x]
  )

# prediction
pred_stack_4_new <- predict(object = pre_stack_4, newX = testset[, x])
```

### 3.3.2 Stacked meta models from scratch

``` r
#Do not include original feature
# put new dataset in the training
pred_stack_5 <-
  stack_ensemble.fit(
    Y = trainset[, y],
    X = trainset[, x],
    newX = testset[, x],
    base_model = c("SL.xgboost", ranger),
    meta_model = c("SL.ranger", "SL.xgboost", "SL.glm"),
    original_feature = FALSE
  )


# predict new dataset separately

pred_stack_6 <-
  stack_ensemble.fit(
    Y = trainset[, y],
    X = trainset[, x],
    base_model = c("SL.xgboost", ranger),
    meta_model = c("SL.ranger", "SL.xgboost", "SL.glm"),
    original_feature = FALSE
  )

pred_stack_6_new <-
  predict(object = pred_stack_6, newX = testset[, x])

#Include original feature
# put new dataset in the training
pred_stack_7 <-
  stack_ensemble.fit(
    Y = trainset[, y],
    X = trainset[, x],
    newX = testset[, x],
    base_model = c("SL.xgboost", ranger),
    meta_model = c("SL.ranger", "SL.xgboost", "SL.glm"),
    original_feature = TRUE
  )

# predict new dataset separately

pred_stack_8 <-
  stack_ensemble.fit(
    Y = trainset[, y],
    X = trainset[, x],
    base_model = c("SL.xgboost", ranger),
    meta_model = c("SL.ranger", "SL.xgboost", "SL.glm"),
    original_feature = TRUE
  )

pred_stack_8_new <-
  predict(object = pred_stack_7, newX = testset[, x])
```

### 3.3.3 Stacked meta models with paralleling computing

``` r
#Do not include original feature
pred_stack_9 <-
  stack_ensemble_parallel(
    object = pred_m3,
    Y = trainset[, y],
    meta_model = c("SL.ranger", "SL.xgboost", "SL.glm"),
    original_feature = FALSE,
    number_cores = 4
  )
pred_stack_9_new <-
  predict(object = pred_stack_9, newX = testset[, x])

#Include original feature
pred_stack_10 <-
  stack_ensemble_parallel(
    object = pred_m3,
    Y = trainset[, y],
    meta_model = c("SL.ranger", "SL.xgboost", "SL.glm"),
    original_feature = TRUE,
    X = trainset[, x],
    number_cores = 4
  )

pred_stack_10_new <-
  predict(object = pred_stack_10, newX = testset[, x])
```

## 4. Cross validation with paralleling computing

Besides the CV for single base/meta model, an ‘external’ CV for the
whole ensemble model is necessary to assess the performance of DEML
modle.

Typically, we conduct a separate holdout sample approach to see the
performance of DEML on unseen data.

### 4.1 Cross validation with base models

``` r
pred_m1_cv <- CV.predictModel_parallel(
  Y = trainset[, y],
  X = trainset[, x],
  base_model = c("SL.xgboost", "SL.ranger"),
  number_cores = 4
)
```

### 4.2 Cross validation with stacked meta models

``` r
#Include original feature
pred_stack_cv_1 <- CV.stack_ensemble_parallel(
  object = pred_m2,
  Y = trainset[, y],
  X = trainset[, x],
  meta_model = c("SL.xgboost", "SL.ranger"),
  original_feature = TRUE,
  number_cores = 4
)

#Do not include original feature
pred_stack_cv_2 <-
  CV.stack_ensemble_parallel(
    object = pred_m1,
    Y = trainset[, y],
    X = trainset[, x],
    meta_model = c("SL.xgboost", "SL.ranger"),
    original_feature = FALSE,
    number_cores = 4
  )
```

## 5. Assess and plot the results

``` r
test_p <-
  assess.plot(pre = pred_stack_10_new$pre_meta$pred, obs = testset[, y])

print(test_p$plot)
```

## Citation

Wenhua Yu, Shanshan Li, Tingting Ye,Rongbin Xu, Jiangning Song, Yuming
Guo (2022) Deep ensemble machine learning framework for the estimation
of PM2.5 concentrations,Environmental health perspectives:
<https://doi.org/10.1289/EHP9752>


NEWS.md
# deeper 1.0.0

* 2022-03-03
1. add paralleling computing for base and meta-model establishment;
2. add predict() to use the established DEML model to predict the unseen new data set;
3. add assess.plot() to replace the assessModel() to assess the relationship between pred and obs;
4. add quantile plot in assess.plot()
5. delete assessModel() since it only works successfully for the model with newX in the training process;
6. delete Method = method.average, since Method parameter was deleted in SL. We can use SL.mean to instead.
7. delete plot_point()
8. add CV.stack_ensemble_parallel() and CV.predictModel_parallel() to conduct the external CV with paralleling
computing operation and return each fold result.


R/predictModel_parallel.R

#' Training and prediction with single model or base ensemble models using paralleling computing
#' @description Training the ensemble models with paralleling computing and make prediction using new data set
#'
#' @param Y the independence, like certain environmental pollutants(eg. PM2.5) in training data that we
#'   want to predict
#' @param X the other predictors(eg. temp, wind speed) in training data that we
#'   want to use to predict air pollutants
#' @param base_model character string in format of "SL.xgboost","SL.randomForest"
#' @param cvControl a list for the control of cross validation, default with 10-fold random selection validRows
#' @param method the method of ensemble the base models, "method.NNLS" (the default).
#' more details see Superlearner
#' @param number_cores the number of CPU cores used in parallel computing
#' @param seed set the numeric seed number,eg. 1234
#' @param ... other parameters that belong to 'SuperLearner' package
#' @return a list including:
#' basemodel: the name of base models
#' model_summary: the summary of base models
#' R2_CV:R-squire
#' RMSE_CV:RMSE
#' base_ensemble_value: the comparison results for base models
#' number_cores: the CPU core
#' Z_matrix: the cross validation results for each base model(Z matrix)
#' time: the modeling time
#'
#' @import SuperLearner
#' @importFrom utils install.packages
#' @examples
#' \dontrun{
#' # Let's predict the PM2.5 values with machine learming methods
#'
#' # Splitting training set into two parts based on outcome: 80% and 20%
#' set.seed(1234)
#' data("envir_example")
#' size <-
#'   caret::createDataPartition(y = envir_example$PM2.5, p = 0.8, list = FALSE)
#' trainset <- envir_example[size, ]
#' testset <- envir_example[-size, ]
#'
#' # set the predictor "PM2.5" and dependence variables
#' y <- c("PM2.5")
#' x <- colnames(envir_example[-c(1, 6)]) # except "date" and "PM2.5"
#'
#' ## conduct the spatial CV
#'
#' # Create a list with 7 folds, each with index of rows to be considered
#'
#' indices <-
#'   CAST::CreateSpacetimeFolds(trainset, spacevar = "code", k = 7)
#'
#' # Rows of validation set on each fold
#'
#' v_raw <- indices$indexOut
#'
#' names(v_raw) <- seq(1:7)
#'
#' pred_m3 <- predictModel_parallel(
#'   Y = trainset[, y],
#'   X = trainset[, x],
#'   base_model = c("SL.xgboost", ranger),
#'   cvControl = list(V = length(v_raw), validRows = v_raw),
#'   number_cores = 4,
#'   seed = 1
#' )
#' ## when number_cores is missing, it will indicate user to set one based on
#' ## the operation system.
#' }
#' @export
predictModel_parallel <- function(Y,
                                  X,
                                  # newX,
                                  base_model,
                                  cvControl = list(),
                                  method = "method.NNLS",
                                  number_cores,
                                  seed = 1,
                                  ...) {
  start_time <- Sys.time()
  if (!requireNamespace("SuperLearner", quietly = FALSE)) {
    warning("need loading required package : SuperLearner",
      call. = FALSE
    )
    utils::install.packages("SuperLearner")
  } else if (requireNamespace("SuperLearner", quietly = TRUE)) {
    if (method == "method.average") {
      method <- do.call(what = .method.average, args = list())
    }
    if (missing(number_cores) == FALSE) {
      number_cores <- number_cores
    } else {
      number_cores <-
        as.numeric(readline(
          prompt = paste0(
            "You have ",
            parallel::detectCores(),
            " cpu cores, How many cpu core you want to use:"
          )
        ))
    }

    while (number_cores > parallel::detectCores()) {
      number_cores <-
        as.numeric(readline(
          prompt = paste0(
            "Your input cores is larger than ",
            parallel::detectCores(),
            ", please set the CPU scores:"
          )
        ))
    }
    if (.Platform$OS.type == "windows") {
      cluster <- parallel::makeCluster(number_cores)
      # Generate seeds for the L'Ecuyer random number generator to create multiple
      # independent random number streams for use in parallel processing nodes
      parallel::clusterSetRNGStream(cl = cluster, iseed = seed)
      # load library 'SuperLearner' to each cluster
      parallel::clusterEvalQ(cluster, library(SuperLearner))
      # copy the function to all clusters
      parallel::clusterExport(cluster, base_model)
      model <- SuperLearner::snowSuperLearner(
        Y = Y,
        X = X,
        SL.library = base_model,
        cluster = cluster,
        cvControl = cvControl,
        method = method,
        ...
      )
      parallel::stopCluster(cluster)
    } else {
      set.seed(seed, "L'Ecuyer-CMRG")
      model <- SuperLearner::mcSuperLearner(
        Y = Y,
        X = X,
        SL.library = base_model,
        cvControl = cvControl,
        method = method,
        ...
      )
    }
  }
  cat("\n\nThe base models Cross validation result:\n\n")
  R2 <- apply(cbind(model$Z, model$SL.predict), 2, caret::R2, obs = Y)
  RMSE <- apply(cbind(model$Z, model$SL.predict), 2, caret::RMSE, obs = Y)
  weight <- c(model$coef, "SL.predict" = 1)
  screen <- rbind(weight = weight, R2, RMSE)
  print(screen)
  Z <- as.data.frame(model$Z)
  names(Z) <- model$libraryNames
  end_time <- Sys.time()
  time <- end_time - start_time

  out <- list(
    # dependence = Y,
    # independence = X,
    # newX = newX,
    basemodel = base_model,
    model_summary = model,
    # ensemble_pre = pred$pred,
    # single_pre = pred$library.predict,
    R2_CV = R2,
    RMSE_CV = RMSE,
    base_ensemble_value = screen,
    number_cores = number_cores,
    Z_matrix = Z,
    time = time
  )

  invisible(out)
}

# the method of average
.method.average <- local(function() {
  out <- list(
    require = NULL,
    computeCoef = function(Z, Y, libraryNames, verbose, obsWeights, ...) {
      # compute cvRisk
      cvRisk <- apply(Z, 2, function(x) mean(obsWeights * (x - Y)^2))
      names(cvRisk) <- libraryNames

      # compute coef
      coef <- rep((1 / ncol(Z)), ncol(Z))
      initCoef <- coef
      initCoef[is.na(initCoef)] <- 0.0
      # normalize so sum(coef) = 1 if possible
      # normalize so sum(coef) = 1 if possible
      if (sum(initCoef) > 0) {
        coef <- initCoef / sum(initCoef)
      } else {
        warning("All algorithms have zero weight", call. = FALSE)
        coef <- initCoef
      }
      out <- list(cvRisk = cvRisk, coef = coef, optimizer = coef)
      return(out)
    },
    computePred = function(predY, coef, ...) {
      if (sum(coef != 0) == 0) {
        stop("All metalearner coefficients are zero, cannot compute prediction.")
      }
      # Restrict crossproduct to learners with non-zero coefficients.
      out <- crossprod(t(predY[, coef != 0, drop = FALSE]), coef[coef != 0])
      return(out)
    }
  )
  invisible(out)
},
envir = .GlobalEnv
) # identify it to the .GlobalEnv
